File: 122 Best Time to Buy and Sell Stock II/main.go
func maxProfit(prices []int) int {
    type State struct {
        idx int
        is_buy bool
    }

    var dp = make(map[State]int)
    var dfs func (i int, is_buy bool) int
    dfs = func (i int, is_buy bool) int {
        if i == len(prices) {
            return 0
        }
        state := State{idx: i, is_buy: is_buy}
        if _, exist := dp[state]; exist {
            return dp[state]
        }
        var sign int
        if is_buy {
            sign = -1
        } else {
            sign = 1
        }
        dp[state] = max(dfs(i+1, is_buy), dfs(i+1, !is_buy) + sign * prices[i])
        return dp[state]

    }
    return dfs(0, true)


    start := 0
    maxProfit := 0
    for end:=1; end < len(prices); end++ {
        if prices[end] > prices[start] {
            maxProfit += prices[end] - prices[start]
        }
        start = end

    }
    return maxProfit
}

File: 122 Best Time to Buy and Sell Stock II/main.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        start = 0
        max_price = 0
        for end in range(1, len(prices)):
            if prices[start] < prices[end]:
                max_price += prices[end] - prices[start]
                start = end
            else:
                start = end
        return max_price
        dp = {}

        def dfs(i, is_buy):
            if i == len(prices):
                return 0
            if (i, is_buy) in dp:
                return dp[(i, is_buy)]
            sign = -1 if is_buy else 1
            dp[(i, is_buy)] = max(
                dfs(i + 1, not is_buy) + sign * prices[i],
                dfs(i + 1, is_buy)
            )
            return dp[(i, is_buy)]

        return dfs(0, True)
